### ISSUE TITLE
Invalid Auth config file

### ISSUE THREAD

**Issue by EndOfMaster** [bmuschko/gradle-docker-plugin#366]
the gradle build:
`import com.bmuschko.gradle.docker.tasks.image.Dockerfile`
`import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage`

`docker {`
   ` url = 'tcp://127.0.0.1:2375'`
   ` registryCredentials {`
       ` url = 'https://registry.cn-beijing.aliyuncs.com'`
        `username = '*********'`
       ` password = '***********'`
`    }`
`}`

`task createDockerfile(type: Dockerfile) {`
   ` destFile = project.file('build/mydockerfile/Dockerfile')`
`    from 'ubuntu:12.04'`
`}`

`task buildImage(type: DockerBuildImage) {`
   ` dependsOn createDockerfile`
 `   inputDir = createDockerfile.destFile.parentFile`
`    tag = 'bmuschko/myimage'`
`}`

the error:
`Failed to parse dockerCfgFile: C:\Users\wzm\.docker\config.json`

the config.json:
`{`
`	"auths": {`
	`	"https://index.docker.io/v1/": {},`
		`"registry.cn-beijing.aliyuncs.com": {}`
	`},`
	`"credsStore": "wincred"`
`}`

**Comment by cdancy** [comment]
@EndOfMaster couple things:

1. Can you post the full stracktrace?
2. Are your registry blocks truly empty or are you just omitting them here? Have you tried without using a `credsStore` to see if things work? I've never used that feature of docker before so am not entirely familiar with it. Might be this is a docker specific problem and nothing to do with the plugin.

**Comment by EndOfMaster** [comment]
I use the docker build command, everything is OK.
this is stracktrace:
https://gist.github.com/EndOfMaster/47af1611886dcf4e71d585eeaec3d1a6

**Comment by orzeh** [comment]
@EndOfMaster the source of the problem is in `docker-java` see [this issue](https://github.com/docker-java/docker-java/issues/806).

**Comment by NickLarsenNZ** [comment]
I'm getting the same thing:
```
* What went wrong:
Execution failed for task ':buildImage'.
> Failed to parse dockerCfgFile: /Users/username/.docker/config.json
```

Other things I've tried which result in the same error:
- An empty ~/.docker/config.json
- A ~/.docker/config.json containing only "{}"

**My ~/.docker/config.json**:
```
{
        "auths": {
                "https://index.docker.io/v1/": {}
        },
        "credsStore": "osxkeychain"
}
```

Interestingly enough, I tried removing only the credsStore key (and preceding comma), which then turns into a NullPointerException with no mention of the config.json file:
```
* What went wrong:
Execution failed for task ':buildImage'.
> java.lang.NullPointerException (no error message)
```

**My build.gradle**:
```
group 'test.jenkinsdocker'
version '1.0-SNAPSHOT'

apply plugin: 'java'
apply plugin: 'war'
apply plugin: 'com.bmuschko.docker-remote-api'

import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage

sourceCompatibility = 1.8

buildscript {
    repositories { jcenter() }
    dependencies {
        classpath 'com.bmuschko:gradle-docker-plugin:3.0.6'
    }
}

repositories {
    mavenCentral()
}

dependencies {
    compile "org.glassfish.jersey.containers:jersey-container-servlet:2.16"
    compile "org.glassfish.jersey.media:jersey-media-json-jackson:2.16"
    testCompile group: 'junit', name: 'junit', version: '4.11'
}

docker {
    url = 'tcp://192.168.99.100:2376'
}

// Seems to fail if  ~/.docker/config.json exists
task buildImage(type: DockerBuildImage) {
    // I have a Dockerfile in ./ which builds successfully if ~/.docker/config.json doesn't exist
    inputDir = project.file('./')
    tag = 'pingpong:gradle'
}
```

**Comment by cdancy** [comment]
@NickLarsenNZ when you removed the `credsStore` did you go back and fill in the actual credentials for the registry?

**Comment by NickLarsenNZ** [comment]
Hi @cdancy, do I didn't think to do that, though shouldn't it be acceptable to not authenticate to the registry (if I'm only pulling)? I don't see why the image build should fail because of a file which isn't so important in that task.

I understand this is a problem with the docker-java project, so I'll jump over there.

Thanks for your reply.

**Comment by cdancy** [comment]
@NickLarsenNZ no I still think you have to put in credentials even for a pull. It may be that some repos are private and locked down and don't want individuals pulling potentially proprietary images.

**Comment by sugun999** [comment]
Hi,

I am having same issue and thows below error. Its unable to pull the image from remote private repository or public repository if its not available locally. Any ideas, what could be the wrong?

*** What went wrong:
Execution failed for task ':presentationlayer:cloud-poc-fast-microservice:buildImage'.
> Failed to parse dockerCfgFile: C:\Users\xxxxx\.docker\config.json**

Here is the excerpt from build.gradle file

`apply plugin: 'com.bmuschko.docker-remote-api'
apply plugin: 'com.bmuschko.docker-java-application'

import com.bmuschko.gradle.docker.tasks.image.Dockerfile
import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage
import com.bmuschko.gradle.docker.DockerRegistryCredentials

def dockerRegistryCredentials = new DockerRegistryCredentials()
        dockerRegistryCredentials.url = 'https://fastregistry.net/v2/'
        dockerRegistryCredentials.username = 'xxxxxxx'
        dockerRegistryCredentials.password = 'xxxxx'
        dockerRegistryCredentials.email = 'xxx.xxx@xx.xxxx.xxx'

task createDockerfile(type: Dockerfile) {
    from "$baseImage"
    maintainer 'Fast team'
}

task buildImage(type: DockerBuildImage) {
    dependsOn createDockerfile
    inputDir = createDockerfile.destFile.parentFile
    registryCredentials = dockerRegistryCredentials
}`

and windows config.json file content is

```
{
	"auths": {
		"fastregistry.net:443": {}
	},
	"credsStore": "wincred"
}
```

The generated dockerfile is:

`FROM fastregistry.net:443/jre8
MAINTAINER Fast`

**Comment by sugun999** [comment]
Hi,

Above issue solved if I specify, dockerRegistryCredentials.url as 'https://fastregistry.net/v2/_catalog'

Thanks

**Comment by cdancy** [comment]
Closing as this appears to be a config issue more than anything else.

**Comment by musketyr** [comment]
I would suggest to reopen this issue as this may break lot of builds. The problem is that Docker for Java is no longer able to read the latest `config.json` format used by Docker - see https://github.com/docker-java/docker-java/issues/806

**Comment by tiny-cababge** [comment]
One more note, you have to delete the config,json in the users/username/.docker folder for this to work.

**Comment by cdancy** [comment]
@tiny-cababge I actually have a mixture of the old and the new way and nothing breaks on my end. It's only when `credsStore` is defined do things pop.

**Comment by tiny-cababge** [comment]
Guess i should rephrase it to change or delete and credstore format from your json?

**Comment by jottinger** [comment]
Could someone please summarize how to get the plugin working again? I've got the same error ("Failed to parse ~/.docker/config.json") and clearing out that file, relogging in to the repos, and running gradle again isn't changing anything at all. Last week I was able to build images; this week I can't. I want to say Docker updated and broke things, but I don't have a good reason to say that yet.

**Comment by cdancy** [comment]
@jottinger you have to remove the `credsStore` section from your config.json file. The recent docker upgrade for me did indeed break things so I suspect it's the same for as well.

**Comment by jottinger** [comment]
@cdancy I removed the `credsStore` and reran `gradle dockerBuildImage --stacktrace` - and got this output:

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':projectName:dockerBuildImage'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:242)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:95)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:235)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:224)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:121)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:77)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:102)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:96)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:612)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:567)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:96)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.NullPointerException
        at com.github.dockerjava.core.AuthConfigFile.decodeAuth(AuthConfigFile.java:162)
        at com.github.dockerjava.core.AuthConfigFile.loadConfig(AuthConfigFile.java:133)
        at com.github.dockerjava.core.DefaultDockerClientConfig.getAuthConfigurations(DefaultDockerClientConfig.java:292)
        at com.github.dockerjava.jaxrs.AbstrDockerCmdExec.getBuildAuthConfigs(AbstrDockerCmdExec.java:38)
        at com.github.dockerjava.jaxrs.BuildImageCmdExec.resourceWithOptionalAuthConfig(BuildImageCmdExec.java:40)
        at com.github.dockerjava.jaxrs.BuildImageCmdExec.callbackNotifier(BuildImageCmdExec.java:118)
        at com.github.dockerjava.jaxrs.BuildImageCmdExec.callbackNotifier(BuildImageCmdExec.java:29)
        at com.github.dockerjava.jaxrs.AbstrAsyncDockerCmdExec.execute(AbstrAsyncDockerCmdExec.java:58)
        at com.github.dockerjava.jaxrs.AbstrAsyncDockerCmdExec.exec(AbstrAsyncDockerCmdExec.java:23)
        at com.github.dockerjava.jaxrs.AbstrAsyncDockerCmdExec.exec(AbstrAsyncDockerCmdExec.java:14)
        at com.github.dockerjava.core.command.AbstrAsyncDockerCmd.exec(AbstrAsyncDockerCmd.java:21)
        at com.github.dockerjava.api.command.AsyncDockerCmd$exec.call(Unknown Source)
        at com.github.dockerjava.api.command.AsyncDockerCmd$exec.call(Unknown Source)
        at com.bmuschko.gradle.docker.tasks.image.DockerBuildImage.runRemoteCommand(DockerBuildImage.groovy:135)
        at com.bmuschko.gradle.docker.tasks.AbstractDockerRemoteApiTask$_runReactiveStream_closure1.doCall(AbstractDockerRemoteApiTask.groovy:57)
        at com.bmuschko.gradle.docker.utils.DockerThreadContextClassLoader.withClasspath(DockerThreadContextClassLoader.groovy:54)
        at com.bmuschko.gradle.docker.utils.ThreadContextClassLoader$withClasspath.call(Unknown Source)
        at com.bmuschko.gradle.docker.tasks.AbstractDockerRemoteApiTask.runInDockerClassPath(AbstractDockerRemoteApiTask.groovy:62)
        at com.bmuschko.gradle.docker.tasks.AbstractDockerRemoteApiTask$runInDockerClassPath.callCurrent(Unknown Source)
        at com.bmuschko.gradle.docker.tasks.AbstractDockerRemoteApiTask.runReactiveStream(AbstractDockerRemoteApiTask.groovy:56)
        at com.bmuschko.gradle.docker.tasks.AbstractReactiveStreamsTask.start(AbstractReactiveStreamsTask.groovy:50)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:141)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:711)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:694)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:95)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 27 more
```

**Comment by jottinger** [comment]
Note that the above stacktrace was with `3.0.6`; I updated to `3.0.8` and got the same results, with slightly different line numbers.

**Comment by cdancy** [comment]
@jottinger I had the same issue and once I massaged my `~/.docker/config.json` file all was Ok. Can you post the content of your file here as I'm curious why it's popping.

I'm also trying to get a new release out so hopefully this will just go away.

**Comment by zinaLacina** [comment]
The issue happens because are not connected to your docker. Connect you, before running the command.
